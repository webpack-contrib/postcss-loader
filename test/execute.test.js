import path from "path";

import {
  compile,
  getCompiler,
  getErrors,
  getCodeFromBundle,
  getWarnings,
} from "./helpers";

describe('"execute" option', () => {
  it('should work with "Boolean" value', async () => {
    const compiler = getCompiler(
      "./jss/exec/index.js",
      {},
      {
        module: {
          rules: [
            {
              test: /style\.(exec\.js|js)$/i,
              use: [
                {
                  loader: require.resolve("./helpers/testLoader"),
                  options: {},
                },
                {
                  loader: path.resolve(__dirname, "../src"),
                  options: {
                    execute: true,
                  },
                },
              ],
            },
          ],
        },
      },
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle("style.exec.js", stats);

    expect(codeFromBundle.css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should work with "postcss-js" parser', async () => {
    const compiler = getCompiler(
      "./jss/postcss-js/index.js",
      {},
      {
        module: {
          rules: [
            {
              test: /style\.js$/i,
              use: [
                {
                  loader: require.resolve("./helpers/testLoader"),
                  options: {},
                },
                {
                  loader: path.resolve(__dirname, "../src"),
                  options: {
                    postcssOptions: {
                      parser: "postcss-js",
                    },
                    execute: true,
                  },
                },
              ],
            },
          ],
        },
      },
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle("style.js", stats);

    expect(codeFromBundle.css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it("should reuse PostCSS AST with JS styles", async () => {
    const spy = jest.fn();
    const compiler = getCompiler(
      "./jss/exec/index.js",
      {},
      {
        module: {
          rules: [
            {
              test: /style\.(exec\.js|js)$/i,
              use: [
                {
                  loader: require.resolve("./helpers/testLoader"),
                  options: {},
                },
                {
                  loader: path.resolve(__dirname, "../src"),
                  options: {
                    execute: true,
                  },
                },
                {
                  loader: require.resolve("./helpers/astLoader"),
                  options: { spy, execute: true },
                },
              ],
            },
          ],
        },
      },
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle("style.exec.js", stats);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(codeFromBundle.css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });
});
