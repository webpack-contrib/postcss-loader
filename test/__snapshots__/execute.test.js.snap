// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"execute" option should reuse PostCSS AST with JS styles: css 1`] = `
"a {
    color: green
}"
`;

exports[`"execute" option should reuse PostCSS AST with JS styles: errors 1`] = `[]`;

exports[`"execute" option should reuse PostCSS AST with JS styles: warnings 1`] = `[]`;

exports[`"execute" option should work with "Boolean" value: css 1`] = `
"a {
    color: green
}"
`;

exports[`"execute" option should work with "Boolean" value: errors 1`] = `[]`;

exports[`"execute" option should work with "Boolean" value: warnings 1`] = `[]`;

exports[`"execute" option should work with "postcss-js" parser: css 1`] = `
"a {
    color: yellow
}"
`;

exports[`"execute" option should work with "postcss-js" parser: errors 1`] = `[]`;

exports[`"execute" option should work with "postcss-js" parser: warnings 1`] = `[]`;
